flink-libraries/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/DynamicResult.java

```diff
@@ -0,0 +1,253 @@
+		return collectTableSink;
+	}
+
+	public void close() {
+		retrievalThread.isRunning = false;
+		retrievalThread.interrupt();
+		monitoringThread.interrupt();
+		collectTableSink.close();
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	private <T> TypedResult<T> handleMissingResult() {
+		// check if the monitoring thread is still there
+		// we need to wait until we know what is going on
+		if (monitoringThread.isAlive()) {
+			return TypedResult.empty();
+		}
+		// the job finished with an exception
+		else if (executionException != null) {
+			throw executionException;
+		}
+		// we assume that a bounded job finished
+		else {
+			return TypedResult.endOfStream();
+		}
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	private class JobMonitoringThread extends Thread {
+
+		@Override
+		public void run() {
+			try {
+				program.run();
+			} catch (SqlExecutionException e) {
+				executionException = e;
+			}
+		}
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	private class ResultRetrievalThread extends Thread {
+
+		public volatile boolean isRunning = true;
+
+		@Override
+		public void run() {
+			try {
+				while (isRunning && collectTableSink.hasNext()) {
+					final Tuple2<Boolean, Row> change = collectTableSink.next();
+
+					if (isChangelog) {
+						processForChangelog(change);
+					} else {
+						processForTable(change);
+					}
+				}
+			} catch (RuntimeException e) {
+				// ignore socket exceptions
+			}
+
+			// no result anymore
+			// either the job is done or an error occurred
+			isRunning = false;
+		}
+
+		private void processForChangelog(Tuple2<Boolean, Row> change) {
+			synchronized (resultLock) {
+				// wait until the last record has been picked up
+				while (isRunning && nextChangeRecord != null) {
+					try {
+						resultLock.wait();
+					} catch (InterruptedException e) {
+						// ignore
+					}
+				}
+				nextChangeRecord = change;
+			}
+		}
+
+		private void processForTable(Tuple2<Boolean, Row> change) {
+			// we track the position of rows for faster access and in order to return consistent
+			// snapshots where new rows are appended at the end
+			synchronized (resultLock) {
+				final List<Integer> positions = rowPositions.get(change.f1);
+				if (change.f0) {
+					// insert
+					materializedTable.add(change.f1);
+					if (positions == null) {
+						// new row
+						final ArrayList<Integer> pos = new ArrayList<>(1);
+						pos.add(materializedTable.size() - 1);
+						rowPositions.put(change.f1, pos);
+					} else {
+						// row exists already
+						positions.add(materializedTable.size() - 1);
+						rowPositions.put(change.f1, positions);
+					}
+				} else {
+					// delete
+					if (positions != null) {
+						// delete row position and row itself
+						final int pos = positions.remove(0);
```

Comment:
Removing from the end is more efficient in ArrayList. Otherwise the content of the backing array needs to be copied.

Reply:
Also the list can be removed from the positions table if it is empty.